opm/autodiff/VFPHelpers.hpp

```diff
@@ -0,0 +1,971 @@
+ */
+inline std::vector<int> commonBlockPattern(
+        const ADB& x1,
+        const ADB& x2,
+        const ADB& x3,
+        const ADB& x4) {
+    std::vector<int> block_pattern;
+
+    extendBlockPattern(x1, block_pattern);
+    extendBlockPattern(x2, block_pattern);
+    extendBlockPattern(x3, block_pattern);
+    extendBlockPattern(x4, block_pattern);
+
+    return block_pattern;
+}
+
+inline std::vector<int> commonBlockPattern(
+        const ADB& x1,
+        const ADB& x2,
+        const ADB& x3,
+        const ADB& x4,
+        const ADB& x5) {
+    std::vector<int> block_pattern = commonBlockPattern(x1, x2, x3, x4);
+    extendBlockPattern(x5, block_pattern);
+
+    return block_pattern;
+}
+
+
+
+
+
+
+
+
+
+
+
+
+/**
+ * Returns the type variable for FLO/GFR/WFR for production tables
+ */
+template <typename TYPE, typename TABLE>
+TYPE getType(const TABLE* table);
+
+template <>
+inline
+VFPProdTable::FLO_TYPE getType(const VFPProdTable* table) {
+    return table->getFloType();
+}
+
+template <>
+inline
+VFPProdTable::WFR_TYPE getType(const VFPProdTable* table) {
+    return table->getWFRType();
+}
+
+template <>
+inline
+VFPProdTable::GFR_TYPE getType(const VFPProdTable* table) {
+    return table->getGFRType();
+}
+
+
+/**
+ * Returns the type variable for FLO for injection tables
+ */
+template <>
+inline
+VFPInjTable::FLO_TYPE getType(const VFPInjTable* table) {
+    return table->getFloType();
+}
+
+
+
+
+/**
+ * Returns the actual ADB for the type of FLO/GFR/WFR type
+ */
+template <typename TYPE>
+ADB getValue(
+        const ADB& aqua,
+        const ADB& liquid,
+        const ADB& vapour, TYPE type);
+
+template <>
+inline
+ADB getValue(
+        const ADB& aqua,
+        const ADB& liquid,
+        const ADB& vapour,
+        VFPProdTable::FLO_TYPE type) {
+    return detail::getFlo(aqua, liquid, vapour, type);
+}
+
+template <>
+inline
+ADB getValue(
+        const ADB& aqua,
+        const ADB& liquid,
+        const ADB& vapour,
+        VFPProdTable::WFR_TYPE type) {
+    return detail::getWFR(aqua, liquid, vapour, type);
+}
+
+template <>
+inline
+ADB getValue(
+        const ADB& aqua,
+        const ADB& liquid,
+        const ADB& vapour,
+        VFPProdTable::GFR_TYPE type) {
+    return detail::getGFR(aqua, liquid, vapour, type);
+}
+
+template <>
+inline
+ADB getValue(
+        const ADB& aqua,
+        const ADB& liquid,
+        const ADB& vapour,
+        VFPInjTable::FLO_TYPE type) {
+    return detail::getFlo(aqua, liquid, vapour, type);
+}
+
+/**
+ * Given m wells and n types of VFP variables (e.g., FLO = {FLO_OIL, FLO_LIQ}
+ * this function combines the n types of ADB objects, so that each of the
+ * m wells gets the right ADB.
+ * @param TYPE Type of variable to return, e.g., FLO_TYPE, WFR_TYPE, GFR_TYPE
+ * @param TABLE Type of table to use, e.g., VFPInjTable, VFPProdTable.
+ */
+template <typename TYPE, typename TABLE>
+ADB gather_vars(const std::vector<const TABLE*>& well_tables,
```

Comment:
Functions and methods should be named with camelCase(). Some of the C codes follow a different scheme though.